### ISSUE TITLE
Out of memory issue after update a browsers to version 104+ (Chrome&Edge)

### ISSUE THREAD

**Issue by MusolQA** [cypress-io/cypress#23391]
### Current behavior

We have configured build to run UI tests on both Edge and Chrome browsers.
Last time I noticed that on Edge test are not executed and finished after 6h (pipeline timeout).
Now we have the same situation on Chrome, that test are not executed.
Without any error/log etc.

What interesting both starting to timeout after update Edge from version 103 to 104 and Chrome from version 103 to 104.
No any code changes between update from 103 to 104, so same code base failing after update to 104.

I run it locally and always have "Out of memory" issue screen during execution. --headed flag added to investigate
I try it on different versions, and always have same issue.

### Desired behavior

Same code should working on browsers version 103 and 104.

### Test code to reproduce

No any code, test are not running, because of out of memory issue.

### Cypress Version

 9.3.1, 9.7.0, 10.0

### Node version

16.16.0

### Operating System

Windows Azure Virtual Machine

**Comment by jsflint007** [comment]
Also seeing the same issue, no major code changes (nothing that would cause any sort of memory leak) and only the updated versioning of browsers. Debugging was always an issue for us with memory crashes but now we are seeing every agent failing to run a test in our azure pipelines due to what we suspect now is a memory issue with the chrome browser being used. Switched to Electron in the same pipeline and we had no issues

**Comment by MusolQA** [comment]
I haven't issues with electron so far. Till update Electron to version 94. After that same issue occurs for all browsers. Super weird.

![image](https://user-images.githubusercontent.com/40713510/185143112-1e0731e0-8a95-4c67-aecf-7e865bbb8bcf.png)

**Comment by vaibhavganatra** [comment]
I'm having the same issue with chrome 104 and chrome 105 beta versions.

**Comment by enriquejortizg** [comment]
I'm having the same issue. What I try to do at the moment is to atomize as much as possible each spec, by making spec change it seems that the memory is "reset" and the problem does not occur, but even if there is only one test case and it is too long, the problem will occur. The most tedious thing is that it doesn't jump any error, it just freezes.

**Comment by astraljames** [comment]
Following up with a conversation on the Cypress Discord server, 
I'm running Chrome 104 with Cypress 9.5.4, running headed, and I'm not having this issue, though others definitely are. The folks who are mentioned that one of them was running 9.6.1, which was receiving the OOM behavior.

My longest spec is 50+ test cases, some of which are individually 50+ lines long for reference, in case spec size is a factor.

If there's any further info I can provide please let me know.

**Comment by richardszanyi-kasa** [comment]
Same issue, chrome 104, same with electron, same like previous versions of cypress or browsers:
- long scenarios takes up the browser's memory and crashes.
- Cypress 9.6 was the same
- Now running Cypress 10.6.0 - same

Tried every option available in github issues such as: 

1. setting chrome memory allocation unsuccessfully
2. numTestsKeepInMemory: 1 - no use

It would be nice if we could config chrome at least to take up more memory from the computer. :(

**Comment by MusolQA** [comment]
@rachelruderman 

Thanks for your answer.
https://github.com/MusolQA/cypress-test-tiny

Here repository with code that causes Out of memory issue.
Please install dependencies, I reproduced in on my local Windows10 machine.

To run please install dependencies (preprocessor to run BDD), and run test.feature file. Below my results.

![image](https://user-images.githubusercontent.com/40713510/185909523-17160a44-b85e-441e-b486-27b7b03e4187.png)


![image](https://user-images.githubusercontent.com/40713510/185909332-2ba7ffce-733a-4c2d-a546-79334750872c.png)

**Comment by MusolQA** [comment]
Btw. I can see that similar issue was reported and solved in January, here: https://github.com/cypress-io/cypress/issues/17853
Seems that its back, maybe it help you where looking for a problem.
I think it should be major priority, cause we are stuck due to issue.

**Comment by vaibhavganatra** [comment]
> Btw. I can see that similar issue was reported and solved in January, here: #17853 Seems that its back, maybe it help you where looking for a problem. I think it should be major priority, cause we are stuck due to issue.

@MusolQA That was reported by me and it's not the same issue. Issue#17853 was caused due to changes in cypress 8.2 and above while this new issue is caused due to changes in memory usage for Chrome 104 and 105 beta

**Comment by robrich7** [comment]
it is due to the new browser versions. I didn't change anything in my code but after the browser update to Chrome 104, many test specs crashed. I downgraded to Chrome 102 and the problems stopped.

**Comment by MusolQA** [comment]
Yes, definitely problem is between cypress and latest chromium relation. However you know, downgrade the browsers version might be workaround for few days, but tests should be run on latest version to be reliable.

**Comment by alecmestroni** [comment]
Same issue here

**Comment by vaibhavganatra** [comment]
See one of the several issues that have been reported on google support for chrome 104 memory.
https://support.google.com/chrome/thread/174889666/increasing-the-memory-limit-on-chrome-v104?hl=en

**Comment by PawelRuszlewicz** [comment]
@rachelruderman any update or progress?

**Comment by robrich7** [comment]
> @rachelruderman any update or progress?

i don't think it's Cypress, but a problem with the Chrome browser, as you can see one post above.

**Comment by PawelRuszlewicz** [comment]
And it's final answer? Due to this chrome change and limitation we cannot run tests with take more 2mins on Chrome/Edge?
As a workaround we can use Firefox that doesn't have this issue with memory

**Comment by enriquejortizg** [comment]
At least Cypress should fail the test case when the memory problem occurs. At this point it just hangs.

**Comment by JessefSpecialisterren** [comment]
Potentially relevant Chromium issue: https://crbug.com/1353749

**Comment by rachelruderman** [comment]
Hi @MusolQA , thank you for creating the [cypress-tiny-test repro](https://github.com/MusolQA/cypress-test-tiny)! I ran it on my personal machine (Mac OS) with Chrome 104 and it passed. Can you tell me a bit more about your Windows Azure Virtual Machine? What OS version are you on and how much memory does it have? TIA!

**Comment by PawelRuszlewicz** [comment]
Also I've same issue on local machine:

![image](https://user-images.githubusercontent.com/101120019/187399144-00cf6848-1911-459d-90e9-acc1c73bef81.png)

**Comment by PawelRuszlewicz** [comment]
Update: After update to Chrome 105,  same issue.

**Comment by alexgil1994** [comment]
Same issue.
Sometimes it crashes with an `out of memory` error, sometimes it just freezes (even on the last step of a running test. the moment it should close, having almost succeeded to finish).
Tried to separate into multiple specs and handle more timeouts/waits, disabling the dom snapshot autosaving to 0, these helped to reduce the problem but it still exists.

Memory related problems initially started months ago after a version update.

**Comment by rszalski** [comment]
We recently upgraded from v9.7 to v10.7 and ran into this issue. This happens in both headless and headed mode. I reproduce this locally on my macOS system but I also see the failures on our CI (Docker w/ node base image).

The repro rate for me is ~50%. It usually happens on the second repeat of a full spec, but if I do not close the crashed window, and instead click "Reload" button it will always crash on subsequent runs. This further corroborates the theory that this is a memory issue unless the process is killed. 

In a spec with 6 tests I'm able to reproduce when running 6 specs, less often when skipping one of them and I wasn't able to reproduce when running 4 or fewer specs. The crash always happens on the 6th step.

<img width="1859" alt="Screenshot 2022-09-01 at 19 23 54" src="https://user-images.githubusercontent.com/184806/187978331-42d010e7-f2eb-4174-85d8-06048a09732d.png">

Chrome v105.0.5195.52 - reproducing reliably
Chrome Canary v107.0.5274.0 - reproducing reliably
Electron v102.0.5005.148 - _not_ reproducing
Firefox v99.0.1 - _not_ reproducing

Are there any Chrome/Cypress debug logs I can enable to be of help here?

**Comment by vaibhavganatra** [comment]
The chrome memory issue will likely be fixed in 106
https://bugs.chromium.org/p/chromium/issues/detail?id=1352649

**Comment by emilyrohrbough** [comment]
FYI: https://github.com/cypress-io/cypress/issues/21135#issuecomment-1235643009

**Comment by PawelRuszlewicz** [comment]
@emilyrohrbough 

I'm just wondering and I just created a same automation test in Playwright for UI Automation, and also use Chrome as a browser to execute and same hardware to run. In this case I don't have any issues with memory, what can suspect that its more Cypress related issue?

**Comment by emilyrohrbough** [comment]
@PawelRuszlewicz Comparing Cypress to Playwright is not 1:1. Playwright does not inject code into the browser to test your page, where as Cypress is injected in the browser when testing your pages -- the reason Cypress can provide Time Travel and dom snapshots for commands that have executed and it's corresponding logs. I have no doubt Cypress is impacting the amount memory used in Chrome, but also wouldn't to say this is the only issue here -- though it could be.

**Comment by enriquejortizg** [comment]
Is any action expected from Cypress?, because this issue is labeled "Needs Reproduction". I understand that the problem comes from the memory reduction they did in Chrome 104, but Cypress being stuck is a Cypress issue, at least it should fail the test case and indicate that it was a memory problem. Currently if someone does not cancel manually, it stays like that forever, which wastes a lot of time

**Comment by PawelRuszlewicz** [comment]
Thats the point. Since 2 weeks our UI test are useless. We don't want split it all smallest scenarios, cause we have E2E wrote in Cypress as well. My managers starting to thinking migrate to the Playwright. 

I've feeling that Cypress doesn't understand how huge impact this issue have for a lot of projects in terms of testing. It should be labeled as higher priority. Instead that, we do not get any clear answer. Where is a problem, and when it will be fixed. Some people say its Chrome issue, later its how Cypress is working under the hood. Something is not clear.

**Comment by henricook** [comment]
+1 extremely frustrating. We've had to keep running old Edge (102) on our CI until this can be fixed, which means we can't test user experience on the latest browser.

**Comment by emilyrohrbough** [comment]
I totally understand the frustration. We are treating this as a high impact issue. Until we can figured out a root cause, we can not provide a clear answer on how and when this will be fixed.

Can you give some general insights on your spec structures? How man commands per test on average? How man tests per spec on average?

**Comment by henricook** [comment]
I can reproduce this running one Spec, with about 300 commands. I'm using the cucumber-preprocessor but I've not mentioned Steps (about 100 steps) or Scenarios (about 4) because I'm aware that's not pure Cypress.

**Comment by StefanLobbenmeier** [comment]
Our test spec is around 900 loc split by 6 cases

**Comment by davidyiuCWI** [comment]
We are also facing this issue in our organization. Any big spec file will consistently crash since a month ago. We have tried many workarounds like numTestsKeptInMemory= 0, moving up to cypress 10.8, the latest node, chrome garbage collection, etc., and none of it work. While this occur less in headless windows, we are still seeing more timeout in our CI

**Comment by rszalski** [comment]
@emilyrohrbough I just tested Cypress v10.8. The issue is still there but the fixes you all did w.r.t. memory consumption did improve it slightly. 

Previously (on v10.7) the pattern of crashes was:

✅ ⛔ ✅ ⛔ ✅ ⛔ ...

Now, on v10.8, we get a bit more "room" and are able to execute 2 full test runs before a crash:

✅ ✅ ⛔ ✅ ✅ ⛔ ✅ ✅ ⛔ ...

 Each emoji is a re-run of a test within the same Cypress/browser session. 🚫 means the test run crashed with "Aw, snap!" (OOM). At that point I restarted the test by clicking the "Reload" that shows next to the Chrome crash page. Of course this is all the same, single spec file. It contains 2 `describe` blocks with 6 and 1 test cases respectively. All share a single `beforeEach` callback. I'd say each has an average of ~65 `cy` commands. `cy.intercept` is being used to await API calls. All this is written in TypeScript.
 
 Not sure how many of the cases reported here are crashes on a single test run vs on re-runs. We experienced both but the crashes on re-runs are reproducing with 100% rate so it is easier for me to test. Hopefully this sheds some light on the root cause. Thank you for working on this!

**Comment by abezzubets** [comment]
If you experience the issue with hanging tests please try disabling the Command Log:
https://docs.cypress.io/guides/references/troubleshooting#Disable-the-Command-Log

It is helped me to solve the issue with hanging tests

**Comment by Tathos20** [comment]
Let me try and give you some additional information. We are using Cypress 9.1.1 and we see Out of Memory with Chrome 104 and 105 after 1 or 2 long tests. If we switch to Electron 94 it is mostly good, but in some rare cases Electron also crashes. 
Upgrading to 10.7.0 resulted to be a bit better, but still Out of memory errors. 
We do not have any workaround, since we have long E2E tests and we can not split them up.

**Comment by PawelRuszlewicz** [comment]
@Tathos20  we have same case, and I'm afraid that we do not get any solution. I reported this issue 17.08, so its over 1 month....

**Comment by Tathos20** [comment]
@PawelRuszlewicz I think we will get a solution, as the problem is spreading out. 
FYI: We are also using windows Azure Windows systems.

**Comment by btimo** [comment]
> Potentially relevant Chromium issue: [crbug.com/1353749](https://crbug.com/1353749)
> 
> Edit: and another one: [crbug.com/1352649](https://crbug.com/1352649)

Thanks for linking those issues - With Chrome 106 beta, the issue is no more present for our use case (106 is getting to stable next week)

**Comment by PawelRuszlewicz** [comment]
@btimo how u run test on beta 106 chrome?

**Comment by PawelRuszlewicz** [comment]
@btimo I still have the issue on 106 beta
![image](https://user-images.githubusercontent.com/101120019/191940003-5d6efb41-86ab-47bd-905a-2e502de2ba37.png)

**Comment by Tathos20** [comment]
I also can reproduce the issue with Chrome 106 beta on Win 10 Pro

**Comment by PawelRuszlewicz** [comment]
Same issue on Chrome 106 beta

**Comment by intoeetive** [comment]
We experience same (or seemingly same) issue on `ubuntu-latest` build with GitHub Actions

**Comment by Tathos20** [comment]
@chrisbreiding  Let me give you another information. I managed to locally reproduce the issue with Chrome, and some small tests. How to reproduce: 
We have a feature file with only three scenarios using (Cucumber preprocessor also)
When the tests finish I don't close the browser, I am just hitting the rerun button. I think at the third or fourth rerun the browser crashed just like when you run a long test.

**Comment by vaibhavganatra** [comment]
There is still a `Needs Reproduction` label on this ticket even though it has been encountered and reproduced by several users here. 
I do understand that this issue is more likely caused by changes in Chrome but cypress memory usage also needs to be improved.

**Comment by PawelRuszlewicz** [comment]
As a owner of ticket I lost my patience... almost 2 months for HIGH priority ticket, without any update, third developer assigned to the this issue. And I'm afraid that even if it is more Chrome related issue, they don't want to fix it because of this Cypress issue. 
We cannot wait more, and have UI tests useless due to this issue and we have to starting migration process to other tool.

I am a little disappointed

**Comment by mjhenkes** [comment]
@PawelRuszlewicz I'm sorry that you haven't had a good experience. I completely agree that cypress has memory issues that we need to resolve. Unfortunately this is a complex issue that is hard to track down. Memory issues could be introduced in Cypress, chrome, or any one of our dependencies. We have been working on this issue and @emilyrohrbough has introduced several changes to try to reduce memory usage.

The reason this issue has changed hands is because it's currently in the midst of our triage process as documented here: https://github.com/cypress-io/cypress/blob/develop/cypress-prioritization-and-triage.md#the-cypress-app-issue-triage

Every two weeks we rotate out the engineers responsible for triaging new issues. With this issue we're still trying to get a local recreation before we can transfer it over to be worked on.

@vaibhavganatra, I understand your frustration. I fully believe this issue to be a legitimate problem. What we are looking for is a recreate of this issue in a repo that we can run so that we can debug against and verify fixes. @emilyrohrbough has implemented several memory fixes, but unless we have a repo demonstrating the problem that we can verify against, they're just guesses and as you can attest they haven't done enough. We can attempt to make reproductions ourselves, but there is not guarantee that we've recreated the same issue you're seeing. And with the high volume of issues we work on we don't always have the time. 

@Tathos20, I'm very excited that you've been able to reproduce this issue locally, could you publish your reproduction to a github repo so that we can work on debugging it?

**Comment by Vladoche** [comment]
I can easily reproduce this problem too since a month. No need of a too long test : above 1min30, my Chrome memory exceeds 2GB and...it crashes randomly. I'm on Cypress 10.8, Chrome V105. BUT ! Since i uninstall each morning my Chrome V105 and rollback to a Chrome V96, there's no more crash all day long. So... i guess Chrome is the real problem. Unfortunately it's just a workaround but it really works !

**Comment by Tathos20** [comment]
@mjhenkes Unfortunately they were existing production tests, so I can not post it on github, but as @Vladoche  mentioned he has some code.

**Comment by PawelRuszlewicz** [comment]
@mjhenkes 

There was a request for repo to reproduce 1 month ago, and it's still available

![image](https://user-images.githubusercontent.com/101120019/192737749-caf8f2d0-b2e3-4a51-8c60-24c6b6e2a54f.png)

https://github.com/MusolQA/cypress-test-tiny


I don't understand how You cannot reproduce it? There are couple of same issues here. A bunch of people have this issue.
Do you try to play on time or make us stupid? Seems that You even didn't read this ticket and previous messages if you cannot see the existing repo with tiny-tests.

I give you tip: clone repository on **Windows** (not Mac as your friend before) and run tests. Error will show up, trust me. Stop answering since 2 months that you cannot reproduce issue, when so many request are on going here.

**Comment by Tathos20** [comment]
Can I suggest a quick workaround? How about adding a new variable into the configuration where we can set after how many tests should cypress close and reopen the browser. With this the tests will be a bit slower, but it can help in some cases? This will not help in that case if there is only one ling test which will die. Also the before all section like login etc. has to be executed again by cypress.

**Comment by emilyrohrbough** [comment]
Has anyone been able to reproduce this without using third-party Cypress plugins? i.e. without the cucumber preprocessor? That's great there is a repo with that plugin, but it'd be best to track down issues with Cypress out-of-the-box.

**Comment by Vladoche** [comment]
Hello Emily ! Yes i reproduced this problem, preprocessor (cucumber) within and without project. In each case, crashes remain as usual. I don't think that any preprocessor is responsible of this kind of problem. And actually, it seems normal, but you know much well than me, i'm not an internal of Cypress. However i finally uninstalled Cucumber Preprocessor because it's not anymore maintained, so...it's a purely personal choice. Woul you really consider the Chrome problem way as a real thing ? I really don't think Cypress is much a pain in this bug.

**Comment by Vladoche** [comment]
You just have to do the same as me : lauch tests with a project with quite long tests (1min30 is a minimum time of execution) with a Chrome version 105 for example. Then, uninstall Chrome, and install instead of it, a v96 (or any version before v100) : you'll see the real way Cypress is supposed to work then. Whatever Cypress version you use, Chrome version is the only crap variable in this kind of bug i think. What do you think then ?

**Comment by MusolQA** [comment]
@emilyrohrbough 

Hold on... You try to tell me that, since almost 2 months, you were not able to just remove preprocessor, and change the feature file for spec,  and reuse existing method, to use just Cypress without any plugins in this reposository to run? It will take my 15 mins, so for you as a dev/owenrs of Cypress it should take no more than 10 mins.  

THAT'S WHY NOBODY SO FAR WAS ABLE TO REPRODUCE ISSUE? REALLY?

What a professionalism...

**Comment by Vladoche** [comment]
Well, man @MusolQA , preprocessor or not, i guess it's not a major piece of this cake of bug. There is NO difference of behaviour between Cypress cucumber preprocessored and one without a preprocessor. I really do think that it's not a point. Don't be too harsh with the cypress team. They may be like anyone of us : simple engineers. Doing their best. My advice to everybody here is simple : stay calm and rollback to V96 of Chrome. Everything will be fine then. Believe me.

**Comment by Vladoche** [comment]
Yep, I agree. And what i think maybe what it is : Cypress.io is not in fault. These recent new versions of chrome introduced a bad way to manage memory. Maybe. But. Just try to rollback to V96 and you'll see by yourself. Uninstall 105. Install V96. You'll see by yourself i repeat. And please stay calm. Nobody here deserve hate because of a misundertanding the real origin of a regression. You know what i mean if you're real testers on field. Stay calm. Stay sharp.

**Comment by MusolQA** [comment]
Sorry for my reaction, maybe it's unnecessary

Maybe yes, maybe not. I don't know cause nobody from Cypress gave resposne here, that's why I am frustrated. No because of this issue, but because of that, after 2 months the answer is that they need a repository to reproduce high priority issue, that could be created in 10 mins for each of Cypress dev... is like a total ignorance.

And you as a tester should know that testing on chrome 96 is not 100% reliable. Try to report a ticket to developer that something is not working on Chrome 96, but manually on 106 is working and wait for a reaction :)

**Comment by vaibhavganatra** [comment]
We all agree that chrome changes are causing this issue. However lets also agree that Cypress shouldn't crash regardless of how chrome behaves. If chrome crashes while test execution then, cypress needs to fail the test and continue to the next test execution. Hanging the entire pipeline due to this is not expected.

**Comment by MusolQA** [comment]
ok, 3  final question to @emilyrohrbough or some from cypress:

We have this issue since at least 2 months, from Chrome 104, however we downgrade to 102/103 and still have this issue so problem could occurred earlier. We are assuming that its because of Chrome update of memory distribution, etc.

1. Does anyone from Cypress contact with Chrome/Edge, got any answer if they will fix it, of something? Now we are going to have 106 (4 versions later) without any fix for that.
2. If not? What is the plan? We are going to use Cypress with a limitation to version 96? You would like to change the current tool behavior 
3. Any chance to solve problems with never finished tests in CI?

**Comment by emilyrohrbough** [comment]
@MusolQA Your negative responses are not helpful and do not align with our [Code of Conduct](https://github.com/cypress-io/cypress/blob/develop/CODE_OF_CONDUCT.md#our-standards) for treating people with respect. Please take a min next time if you are frustrated so we can have constructive conversations to move forward with solving this issue.

As both @mjhenkes and I have expressed, we are taking this issue seriously and understand this is a real problem for many users. Reproductions are a great way for us to start to identify root-causes of issues and one benefit of being an open source project is the input from the community for both finding and solving these important issues. We **are** looking into this and it does take time, and the root-cause isn't obvious or we'd have fixed it! 😄 

We'd love if you'd continue to respond with information that will help solve the issue at hand if you are able to dig into it yourself while you are waiting for a fix..

**Comment by MusolQA** [comment]
@emilyrohrbough I agree that my previous messages were out of topic and I apologize for that, so I removed those. I cannot agree that those do not align with Code of Conduct, what more, those are aligned with point "Gracefully accepting constructive criticism". Especially that I didn't offend any people here and I have respect for all people, 
I didn't point my negative response to any person, just criticized the overall situation regarding the issue...

I said aloud what a lot of people think, that asking for a repository after 2 months since of ticket creation to **reproduce issue** is not it is not commensurate with treating it as a higher priority, that's my feeling and I can express them.

Anyway, we should stop discuss out of topic. Good to hear that you try to investigate the issue. Wish that it will be solved quickly, cause I can see that problem is started to be widespread. Big request to send the update to us and thanks for last response and all your work so far. Good that solving approach is ongoing, cause this silent and no any response ...it caused our uncertainty and I am asking for your understanding too, Thanks.

**Comment by jweingarten** [comment]
Chrome 106.0.5249.61 does not fix the problem for me. The same test runs fine on Chrome 103 on my Mac, but with 106 it runs out of memory. 
Setting `numTestsKeptInMemory` to 1 is not helping. Setting it to 0 worked a few times, but not always.
Setting `CYPRESS_NO_COMMAND_LOG=1` I find like a not acceptable solution. Yes I can get the command logs into a file, but that feature is part of the attraction of Cypress. :-) 

```
  cypress:launcher:browsers chrome stderr: <--- Last few GCs --->

[83155:0x13800864000]   279299 ms: Scavenge 935.9 (991.0) -> 931.4 (993.2) MB, 6.5 / 0.0 ms  (average mu = 0.998, current mu = 0.998) task; 
[83155:0x13800864000]   279737 ms: Scavenge 937.8 (993.4) -> 931.7 (994.5) MB, 3.4 / 0.0 ms  (average mu = 0.998, current mu = 0.998) allocation failure; 
[83155:0x13800864000]   279995 ms: Scavenge 938.4 (994.7) -> 933.3 (994.7) MB, 4.6 / 0.0 ms  (average mu = 0.998, current mu = 0.998) task; 


<--- JS stacktrace --->

[83155:259:0929/140255.262562:ERROR:v8_initializer.cc(730)] V8 process OOM (Oilpan: Reserving memory.). +5m
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 50601 } +2s
  cypress:server:socket-base socket-disconnecting transport close +370ms
  cypress:server:socket-base socket-disconnect transport close +0ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 50592 } +1ms
  cypress:server:request aborting { requestId: 'request2704' } +499ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51637 } +1ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51650 } +1s
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51644 } +82ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51655 } +456ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51619 } +9ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51662 } +427ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51659 } +41ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51668 } +25ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51665 } +399ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51677 } +1s
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51674 } +27ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51671 } +634ms
  cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 51680 } +33ms
  cypress:server:util:process_profiler current & mean memory and CPU usage by process group:
  cypress:server:util:process_profiler ┌─────────┬───────────────────┬──────────────┬────────────────────────────────────────────┬────────────┬────────────────┬──────────┬──────────────┬─────────────┐
  cypress:server:util:process_profiler │ (index) │       group       │ processCount │                    pids                    │ cpuPercent │ meanCpuPercent │ memRssMb │ meanMemRssMb │ maxMemRssMb │
  cypress:server:util:process_profiler ├─────────┼───────────────────┼──────────────┼────────────────────────────────────────────┼────────────┼────────────────┼──────────┼──────────────┼─────────────┤
  cypress:server:util:process_profiler │    0    │     'Chrome'      │      6       │ '83112, 83215, 83214, 83146, 83142, 83141' │     0      │     96.25      │   266    │    296.88    │   336.97    │
  cypress:server:util:process_profiler │    1    │     'cypress'     │      1       │                  '82474'                   │     0      │     23.62      │  187.51  │    191.16    │   209.13    │
  cypress:server:util:process_profiler │    2    │     'plugin'      │      1       │                  '82486'                   │     0      │       0        │  54.34   │    94.64     │   268.81    │
  cypress:server:util:process_profiler │    3    │    'launchpad'    │      1       │                  '82658'                   │     0      │      0.33      │  53.63   │     84.7     │    97.85    │
  cypress:server:util:process_profiler │    4    │ 'electron-shared' │      2       │               '82479, 82537'               │     0      │      0.31      │  39.28   │    56.01     │    61.57    │
  cypress:server:util:process_profiler │    5    │      'other'      │      2       │               '84250, 84251'               │     0      │      0.51      │   2.57   │     2.8      │    5.86     │
  cypress:server:util:process_profiler │    6    │      'TOTAL'      │      13      │                    '-'                     │     0      │     117.8      │  603.32  │    708.44    │    929.8    │
  cypress:server:util:process_profiler └─────────┴───────────────────┴──────────────┴────────────────────────────────────────────┴────────────┴────────────────┴──────────┴──────────────┴─────────────┘ +10s

```

**Comment by vaibhavganatra** [comment]
Tried running my pipeline on chrome 107 beta (107.0.5304.18) and it went through successfully without hanging. I would suggest you all try executing tests on 107 beta.

**Comment by ksmills9** [comment]
Hi all,

While the Cypress team looks for a fix, I've found a workaround.

The only solution I've been able to come up with lately is to split up the long test suites into smaller sections.
So I took my 10 tests in one file, and split them up into 3, 3 and 4 tests in 3 different files. Not ideal, but works so far while there is not yet a solution. Very much subjective and trial and error though.

Hope it helps

**Comment by Vladoche** [comment]
@ksmills9 Yep ! It works well ! But right now the solution is simply to test on the next relase 107 of chrome (beta version incoming). And all the crashes are over. We now see the end of this nightmare. Or i hope so...

**Comment by sdaconceicao** [comment]
Thanks @vaibhavganatra, I just installed the chrome 107 beta and can also confirm I'm no longer dealing with constant crashes.

**Comment by Tathos20** [comment]
I can confirm that chrome 107 solved the issue to me. So we are waiting to be released.

**Comment by mjhenkes** [comment]
Update: I am able to (mostly) reliably reproduce this issue in run mode on my mac (i9, 16 gb ram) against chrome 105 with the following test: 

```javascript
it('runs out of memory on a single test', () => {
  for (let index = 0; index < 500; index++) {
    cy.log(`attempt #: ${index}`) // fails at about attempt 337
    cy.visit('https://rahulshettyacademy.com/AutomationPractice/')
    cy.get('#autocomplete')
    .should(
      'have.attr',
      'placeholder',
      'Type to Select Countries',
    )
  }
})

```

If I modify the test to have 500 it blocks with one call each the test will pass.

```javascript
for (let index = 0; index < 500; index++) {
  it.only(`runs out of memory with 500 tests: ${index}`, () => {
    cy.visit('https://rahulshettyacademy.com/AutomationPractice/')
    cy.get('#autocomplete')
    .should(
      'have.attr',
      'placeholder',
      'Type to Select Countries',
    )
  })
}
```

additionally if we run the first test with the command log disabled (`CYPRESS_NO_COMMAND_LOG=1`) the test also succeeds. My theory is that there is at least one memory leak in the command log related to long tests that is made worse in chrome 100 - 106 for some reason. My guess is that we're storing dom snapshots in run mode when we shouldn't be. But I need to eliminate more variables to be sure. 

I don't think that this is the only scenario that causes chrome to crash but this is definately one of them and i'm glad to hear chrome 107 at least handles the problem better.

**Comment by ursmeili** [comment]
Chrome 107 beta seems to really solve the issue, in a sense it doesn't crash anymore. However, it still uses an awful lot of memory. Running a certain (admittedly really large) cypress e2e test sometimes lets chrome 107 beta use over 4 GB of memory (according to the Windows Task Manager). With Chrome 106, the "out of memory" error already came when chrome.exe used around 2.5 GB of memory. So it seems they didn't really fix the memory leak, they just increased some threshold...

**Comment by PawelRuszlewicz** [comment]
Yep, can confirm that 107:beta it didn't crashed at least ! :)

Bad news that it still consume ~80-90% of my 16GB memory, and whole Cypress starting working very slow in long E2E scenarios (for example typing text to the input is laggy etc.)

@mjhenkes  Good to hear that you find one of the problem!

**Comment by chrisbreamewoodmac** [comment]
I can see that a lot of people are suggesting that it's Chrome which is at fault here, but I'm not running any of my tests on Chrome at all, and I still see the issue. I also noticed that some people have suggested Firefox is not affected (or affects as much), but I can't verify this as Firefox doesn't launch for me for some reason.

Anyway, my setup:
* Cypress v10.9.0
* Electron v102 / Edge v106
* cucumber preprocessor v4.3.1
* Windows 10 x64
* 32GB RAM

Steps:
1. Open Cypress in headed/UI mode
2. Open task manager and watch Cypress RAM usage - currently sits at around 350MB
3. Kick off one .feature (consists of 4 tests) and watch as the RAM usage hits around 5GB (no crash yet, as I have tons of RAM and I'm on a 64 bit Windows)
4. Watch task manager and wait about 30 seconds. You can see RAM usage drops to around 4.9GB through some sort of garbage collection
5. Leave the Cypress window open and just restart the same test again (by hitting "r" on the keyboard) - RAM usage climbs up to around 5.5GB and then the Cypress window becomes unresponsive. CPU/GPU usage falls to 0% and RAM goes down to around 350MB again
6. Console states "We detected that the Chromium Renderer process just crashed."

Closing and reopening the Cypress window will allow me to run that test again (once) just fine. In fact, I can run any test once! But if I re-run the same test more than once, then I **always** exceed some sort of RAM threshold and it dies.

The same issue appears to affect Edge as well. The point I'm trying to make: it's not just Chrome, it looks like it affects everything that is Chromium based

**Comment by chrisbreamewoodmac** [comment]
Oh, and of note:
* I tried this with Cypress 10.9.0 and the built-in version of Electron, and it ran out of RAM in UI/headed mode
* I tried this with Cypress 10.9.0 and Edge 106, and it ran out of RAM in UI/headed mode
* I tried this with Cypress 10.8.0 and the built-in version of Electron, and it ran out of RAM in UI/headed mode
* I tried this with Cypress 10.8.0 and Edge 106, and it ran out of RAM in UI/headed mode
* I tried this with Cypress 8.0.0 and the built-in version of Electron, and **it ran perfectly in UI/headed mode**
* I tried this with Cypress 8.0.0 and Edge 106, and **it ran perfectly in UI/headed mode**
* I tried this with Cypress 10.9.0 and the built-in version of Electron (102), and I was able to run the entire suite (that's 18 .feature files, each with between 1-20 tests in) without running out of RAM in headless mode

This leads me to believe that it is not Chromium at all, but something to do with Cypress itself. Same browser version, but different Cypress version yields a massive difference in RAM usage.

**Comment by davidyiuCWI** [comment]
I can also confirm that using Chrome 107 beta also fixes the Chrome crash in my testing. We will wait until chromium 107 is in production to run our full test.

**Comment by nickpm21** [comment]
This has been a consistent issue for my tests as well, 10.8.0 Cypress and Chrome v106. The thing is I have 32 GB of RAM and from the view in my Task Manager it doesn't look like it's even getting close to 100% usage but will still trip the Aw Snap! Out of Memory error. So I'm wondering what's the threshold that's making Cypress trigger this.

**Comment by henricook** [comment]
In the name of the many people subscribed to this thread can I ask you please to only comment if you're still experiencing this issue in Chrome 107 beta - otherwise I think it's safe to say this issue will be fixed once Chrome 107 is released.

**Comment by PawelRuszlewicz** [comment]
@henricook I agree that we shouldn't spam this ticket. However I cannot agree that Chrome 107 will solve the issue. It gonna fix it partly. On beta version we can still huge memory usage of Cypress and in case of long E2E scenarios the overall execution starting to work slower. Also you cannot guarantee that any of next Chromium version will not cause this issue again.

In my opinion, we should not rely on the next version of Chrome, but eliminate the root of the problem...

**Comment by Daniel-Fig** [comment]
Sorry everyone, but beta 107 increased how long it takes to crash (it reach the mark of 330 seconds before crash), but it is still crashing
![Screen Shot 2022-10-14 at 2 29 16 PM](https://user-images.githubusercontent.com/103145558/195906840-36761f4c-074b-49b6-bdce-d0f86bba1c58.png)

**Comment by CoryDanielson** [comment]
@chrisbreamewoodmac you may want to start a new issue for your crash. It sounds different.

The issue at the root of this thread is specific to Chrome 103 up to the current versions. Some have worked around the issue by running an older version of Chrome for their tests.

https://github.com/cypress-io/cypress/issues/23391#issuecomment-1246700332
> +1 extremely frustrating. We've had to keep running old Edge (102) on our CI until this can be fixed, which means we can't test user experience on the latest browser.


-----

I just updated to Chrome 106 today, and it's still crashing but I think it's improved a bit? When I run my tests without garbage collection, they crash after 100 seconds. When I run them [with garbage collection](https://github.com/cypress-io/cypress/issues/8525#issuecomment-1232991216), they crash after ~600 seconds (Chrome 106.0.5249.119 - Cypress 9.7.0) I think this crashed for me a while back at 3-400 seconds.

**Comment by Tathos20** [comment]
Upgrading to cypress 10.10.0 made things worse! 
Electron and Chrome is constantly crashing with out of memory.
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/114226068/196410728-ff68c7d4-d36f-45f2-a538-101524ca1ff6.png)

**Comment by Tathos20** [comment]
Meanwhile Chrome usually takes about 2.5gb of memory during test execution, but when there are some heavy assertions it goes up a lot. At about 6GB it will crash.

**Comment by Tathos20** [comment]
Meanwhile Chrome usually takes about 2.5gb of memory during test execution, but when there are some heavy assertions it goes up a lot, At about 6GB it will crash.

**Comment by AtofStryker** [comment]
I have been able to reproduce in chrome 106 but it has been incredibly intermittent and takes some time with the 500+ test cases/visit. Running the profiler doesn't show an increase in memory usage over time, but I need to also measure the process

**Comment by wctiger** [comment]
Not sure if this helps isolate the root cause. I am running an older version of Cypress (v6.4.0) with Chrome 105 in headed mode and I can reproduce the crash consistently. Meanwhile I don't see the same failures when running headless.

**Comment by idanElitzur** [comment]
Any news regarding these issues?
I've got them as well, browser sending huge of network request amount in the background,
it doesn't matter which browser type is execute the tests.
the browser ls stuck and crash after the cpuPrecent is reach up to 250
with 7-8 running processes.

cypress: 10.10.0
node: 14.19.1
chrome: 107
mac os: 12 

@chrisbreiding 
@ryanthemanuel 
please your help! :pray:

**Comment by gunslingor** [comment]
Chrome 107.0.5304.107 (Official Build) (64-bit) - I get this after running 50% of my tests. I got 32GB of RAM on this.

![image](https://user-images.githubusercontent.com/11238428/201759767-64430ccc-b5bf-4ab1-9841-c64755eceee0.png)

**Comment by hegeds** [comment]
I've done some digging and just for everybody: currently there is a 4GB memory limit for tabs in chromium. https://bugs.chromium.org/p/chromium/issues/detail?id=416284

so after some point it doesn't matter how much memory you have, if the browser tab reaches that limit it will crash.

**Comment by PawelRuszlewicz** [comment]
I just forgot about that reported by me issue. I already migrated to Playwright. And seems that was good decision, especially that since August we don't have any solution, for this high priority blocking many companies issue.

Seems that Cypress is good tool, but not for long E2E scenarios, due to this Chrome limitation.

**Comment by ryanthemanuel** [comment]
Thank you @hegeds and others for the additional information. We are continuing to try and diagnose this in-house. We've had some sporadic success at recreating it, but are still trying to determine the exact root cause.

**Comment by pgcayen** [comment]
It's not particularly relevant to cypress, but I bumped into this thread while trying to debug my Microsoft Teams tab running out of memory at least twice per day. This wasn't an issue for me until I upgraded chrome beyond 103

The 4GB memory limit is just insufferable :sob:

**Comment by emreSmrt** [comment]
The same issue 
CYPRESS 11.1
CHROME 107 
<img width="905" alt="Screenshot 2022-11-17 at 09 00 35" src="https://user-images.githubusercontent.com/83953275/202390097-bee2b0f5-76ff-48ba-b585-d657a527ba77.png">

<img width="840" alt="Screenshot 2022-11-17 at 09 02 35" src="https://user-images.githubusercontent.com/83953275/202390079-908eae16-61b8-42b2-92f3-382c5cb3e64e.png">

**Comment by Marcelinna** [comment]
I have the same problem 10.10.0, chrome 107

**Comment by chrisbreamewoodmac** [comment]
> @chrisbreamewoodmac you may want to start a new issue for your crash. It sounds different.
> 
> The issue at the root of this thread is specific to Chrome 103 up to the current versions. Some have worked around the issue by running an older version of Chrome for their tests.
> 
> [#23391 (comment)](https://github.com/cypress-io/cypress/issues/23391#issuecomment-1246700332)
> 
> > +1 extremely frustrating. We've had to keep running old Edge (102) on our CI until this can be fixed, which means we can't test user experience on the latest browser.
> 
> I just updated to Chrome 106 today, and it's still crashing but I think it's improved a bit? When I run my tests without garbage collection, they crash after 100 seconds. When I run them [with garbage collection](https://github.com/cypress-io/cypress/issues/8525#issuecomment-1232991216), they crash after ~600 seconds (Chrome 106.0.5249.119 - Cypress 9.7.0) I think this crashed for me a while back at 3-400 seconds.

I really don't believe that this is Chrome-specific (or rather that it's Chrome v103-specific), nor do I think it's a separate issue to what everybody on this issue is reporting. As far as I can tell, using Cypress v10.11.0, my tests regularly go over the RAM limit and crash when running headed on:
Electron v106 (version which is built in)
Chrome 107
Edge 107

If I run headless, they will not crash. This did not happen with the **exact same tests** on an earlier version of Cypress.

Repro steps:
1. Open Cypress 10.11.0 headed.
2. Run a test
3. Watch as the RAM usage climbs and climbs
4. If it didn't crash first time, simply re-run the same test again

Surely there is enough people now telling you that this is a problem that you will treat it as a valid problem?

**Comment by CoryDanielson** [comment]
My bad @chrisbreamewoodmac - your reproduction steps were the only to mention rerunning a test within this thread. I'm also sure that you're experiencing the same issue.

Regarding
> Surely there is enough people now telling you that this is a problem that you will treat it as a valid problem?

I do not work for Cypress. The cypress team has been working on this but it seems like Chrome pulled the rug out from under them without notice here.

I'm having the same issue at work as well, but we're just working around it, and using `cypress run` during our automated test runs which has no issue.

**Comment by CoryDanielson** [comment]
> Thank you @hegeds and others for the additional information. We are continuing to try and diagnose this in-house. We've had some sporadic success at recreating it, but are still trying to determine the exact root cause.

Do you need access to a large web app to reproduce this more consistently? The app that I work on is so large and complex that you could probably reproduce the issue by navigating around in our app during a free trial period.

**Comment by ErickFlores13** [comment]
I had the same problem, idk why but i added a  cy.clearCookies() before each test case and it fixed my problem, maybe this works for that type of tests

**Comment by mschile** [comment]
All, we are continuing to investigate this issue and I logged a [Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1392601) to hopefully understand what changed between v103 and v104+.

In the meantime, you can try this [suggestion](https://github.com/cypress-io/cypress/issues/8525#issue-696149834) by @CoryDanielson to force garbage collection.

**Comment by mschile** [comment]
Some other configuration options that may help with memory:

* Reducing the [`numTestsKeptInMemory`](https://docs.cypress.io/guides/references/legacy-configuration#Global)
* Disabling the [command log](https://docs.cypress.io/guides/references/troubleshooting#Disable-the-Command-Log)

**Comment by CoryDanielson** [comment]
@mschile if you read through some of the chrome bugs linked in this GitHub issue, you'll find your way to this one. You can follow the related bug that it was merged into, too. This issue is related to a reduction in oilpan memory capacity

https://bugs.chromium.org/p/chromium/issues/detail?id=1352649

**Edit**: is the command log written using a `<table>` element? This bug could be related, if so https://bugs.chromium.org/p/chromium/issues/detail?id=1359443
- **Edit**: Command log uses lists.

**Comment by CoryDanielson** [comment]
Anecdotal evidence from this morning:

**Cypress 11.2.0, Chrome 107.0.5304.110 (Official Build) (x86_64)**
All of these runs were with the command log enabled.
- **crashed** `numTestsKeptInMemory: 1, no gc` OOM on the 5th test. Memory ran up to 5GB and crashed.
- **crashed** `numTestsKeptInMemory: 1, with window.gc() calls`. Memory >2gb after 7 tests. 2.8gb after 10. OOM crash soon after
- **suite passed** `numTestsKeptInMemory: 0, no gc` the suite of 31 tests passed. I didn't keep an eye on the memory for this run. finished at 1gb
- **suite passed** `numTestsKeptInMemory: 0, with window.gc() calls` running great. memory gc'd down to 800mb during the 17/18th test. suite finished after 31 tests & 714mb memory.

Seems like an issue with snapshots. `numTestsKeptInMemory: 0` is a good workaround for now.

**Comment by mschile** [comment]
Thanks @CoryDanielson, I would also be curious what happens on Chromium 103. If you have time, would you mind running against 103? You can download previous versions here: https://chromium.cypress.io/

**Comment by hegeds** [comment]
@CoryDanielson I have to disagree unfortunately, setting the `numTestsKeptInMemory` will only help a little (especially when we are seeing the problem in the pipeline, where the value of the `numTestsKeptInMemory` is forced to be 0). The same goes for calling the garbage collector. These are just bandaids on an open wound.

These solution might help someone just enough, that their current test will pass, but in the current state cypress unable to handle complex test runs. I've created a snippet with which one could easily create something to reproduce the problem

```
describe('empty spec', () => {
  for (let index = 0; index < 100; index++) {
      it('passes', () => {
        for (let j = 0; j < 10; j++) {
            cy.task('log', j);

            cy.visit('**your-test-site**')
            cy.wait(500);

            cy.get('**navbar-in-the-side**').click();
            cy.contains('**some-element-to-verify-the-navigation-is-successfull**').should('exist');
            cy.wait(500);

            cy.get('**navbar-in-the-side**').click();
            cy.contains('**some-element-to-verify-the-navigation-is-successfull**').should('exist');
            cy.wait(2000);
        }
    })
    }

    afterEach(() => {
      cy.window().then(win => {
          win.gc();
          win.gc();
          win.gc();
          win.gc();
          win.gc();
      });
    });
})
```

with this setup I see the following memory usage:
```
PID    COMMAND          MEM
56312  Google Chrome He 604M
56312  Google Chrome He 614M
56312  Google Chrome He 677M
56312  Google Chrome He 747M
56312  Google Chrome He 755M
56312  Google Chrome He 799M
56312  Google Chrome He 835M
56312  Google Chrome He 836M
56312  Google Chrome He 836M
56312  Google Chrome He 836M
56312  Google Chrome He 839M
56312  Google Chrome He 839M
56312  Google Chrome He 841M
56312  Google Chrome He 887M
56312  Google Chrome He 947M
56312  Google Chrome He 965M
56312  Google Chrome He 991M
56312  Google Chrome He 992M
56312  Google Chrome He 993M
56312  Google Chrome He 993M
56312  Google Chrome He 994M
56312  Google Chrome He 995M
56312  Google Chrome He 995M
56312  Google Chrome He 997M
56312  Google Chrome He 1044M
...
56312  Google Chrome He 1750M
56312  Google Chrome He 1750M
56312  Google Chrome He 1750M
56312  Google Chrome He 1750M
56312  Google Chrome He 1752M
56312  Google Chrome He 1752M
56312  Google Chrome He 1753M
```

I also used Cypress 11.2.0, Chrome 107 to produce these results.

I think this issue existed way before, and the memory constraints in the new chrome release just surfaced these problems for more people.

**Comment by CoryDanielson** [comment]
@hegeds If you're running in a pipeline why are you even using `cypress open`? This issue is specific to cypress open.

I'd argue that the codebase I work on has "complex" tests. If I reached an OOM after 5 `it` blocks, it's a pretty decent indicator of how massive the app is that we're testing against. Running with `numTestsKeptInMemory: 0` allowed me to run 31 tests without OOM and low memory usage.

I agree that setting numTestsKeptInMemory to 0 is a bandaid, but imo you shouldn't be running massive suites of tests with `cypress open`. Cypress crashing often due to OOM is annoying, but you can work around it for a while.

---

@mschile I don't have time to debug it further. Running up against holiday time off.

**Comment by hegeds** [comment]
@CoryDanielson sorry for not being specific, but I execute `cypress run` and I'm seeing this issue there as well. The result I just posted was also executed with `cypress run`.

**Comment by CoryDanielson** [comment]
@hegeds do you get OOM issues with `cypress run`?

**Comment by hegeds** [comment]
@CoryDanielson yes

**Comment by CoryDanielson** [comment]
Honestly you might want to open a separate issue for that - the conversation could drown everything else here. I'm running `cypress run` now across like 190 test files, and what I see is that the Chrome Helper process gets killed and a new one is spun up between test files.

We avoid this issue where I work by keeping our test files somewhat small, so that they can be parallelized easier too.

**Comment by hegeds** [comment]
but that's my point, that they are the same issue, for some reason there is a memory leak in the cypress process. Most people saw the problem when running with cypress open, having snapshot stored etc because those require extra memory, and with those you can more easily get over the limit, but the underlying issue can be seen running the test in headless mode.

I know, that I could separate the tests into more files, and then the helper is cleared up, but this is not a solution, this a work around for the memory leak. At the current state cypress unable to handle complex test scenarios.

**Comment by richardszanyi-kasa** [comment]
@CoryDanielson I agree with @hegeds that this might be the same memory problem both with `cypress run` and `cypress open`.

We were seeing failures due to that in local runs as well as CI runs. (open, and run) Of course, the latter is less frequent. 
One thing I noticed, this is directly proportional to the HTML elements cypress gets for each `cy.get()`.. The more we wanted to "find" an element and do something with it, the more likely it caused a crash in the browser. 

Of course, in CI every test file opens up a new browser instance, while locally, you are basically in the same session.. 

I think it has something to do with the `numTestsKeepInMemory` not working properly because in my mind it should avoid these errors exactly by clearing any "pollution" between tests.

The error might came up now more and more because of two things: 

- Chromium has changed in terms of memory usage, as far as I know they only allow 4GB of ram / tab. (I'm not sure when it occured)
- People started to use Cypress more and more for bigger E2E testing, and test frameworks started to become scaled up due to that.
- If this doesn't get wrapped around sooner or later I suspect that many people who use cypress for E2E testing might migrate their code to different tools..

**Comment by mschile** [comment]
@MusolQA, @richardszanyi-kasa, @hegeds, and any others experiencing this issue using `cypress run`, could you please try installing the below version of Cypress to see if your memory issues are fixed:

```
npm install https://cdn.cypress.io/beta/npm/12.1.0/linux-x64/mschile/chrome_memory-5ccc7c6ae5d2a31e9625148db450b53719b1ad53/cypress.tgz
```

Note: this version is based off Cypress 12.1.0

**Comment by nickpm21** [comment]
This is getting pretty egregious the tests are running OOM barely a minute in

**Comment by brian-mann** [comment]
@nickpm21 are you running in `cypress open` or `cypress run`? can you try matt's approach above your comment here? 

https://github.com/cypress-io/cypress/issues/23391#issuecomment-1351858785

**Comment by nickpm21** [comment]
> @nickpm21 are you running in `cypress open` or `cypress run`? can you try matt's approach above your comment here?
> 
> [#23391 (comment)](https://github.com/cypress-io/cypress/issues/23391#issuecomment-1351858785)

cypress open. What happens is cypress opens and can run a test fine the first time, but if you rerun the test it will fail. Cypress opens with about 1 GB of RAM usage and after the first test runs it keeps a steady state of 2.5 GB. During the first test Cypress' Chrome browser will go from 1 GB all the way up to around 5 GB. But because the return state is around 2.5 GB it will fail because the RAM usage will go much higher.

**Comment by mschile** [comment]
👋 @nickpm21, thanks for the additional details. By default, Cypress will store snapshots and command details (which are useful for debugging) from the last 50 tests in memory. Depending on several factors, such as the size of your tests, the size of the app be loaded, what commands are being run, and so forth, the amount of memory consumed can be quite large. Thankfully, you can control this behavior with the [numTestsKeptInMemory](https://docs.cypress.io/guides/references/legacy-configuration#Global) config option.

Could you try using the Cypress version I provided [above](https://github.com/cypress-io/cypress/issues/23391#issuecomment-1351858785) and setting the `numTestsKeptInMemory` to 5 and see if that resolves your memory issues?

**Comment by richardszanyi-kasa** [comment]
@mschile As many of us noted previously in this thread - ` numTestsKeptInMemory` is only a minor possible workaround.. 
For us, it didn't resolve anything, so we had to force decoupling of some of the test suites that was a memory hog. 

It is really unfortunate because cypress with this memory issue is currently in no shape to test E2E on a bigger project. We sometimes joke about migrating to Selenium 4, I hope it stays an inside joke.

**Comment by henricook** [comment]
> @mschile As many of us noted previously in this thread - ` numTestsKeptInMemory` is only a minor possible workaround.. For us, it didn't resolve anything, so we had to force decoupling of some of the test suites that was a memory hog.
> 
> It is really unfortunate because cypress with this memory issue is currently in no shape to test E2E on a bigger project. We sometimes joke about migrating to Selenium 4, I hope it stays an inside joke.

😂  We just migrated from Selenium to Cypress and are rue'ing the day - fingers crossed that someone has a chance to work on this soon. It's been a long long wait for this to get fixed.

**Comment by JessefSpecialisterren** [comment]
@mschile Could you also release a Windows version of that special build? Our project is very Windows-based 😇

**Comment by mschile** [comment]
@JessefSpecialisterren, the above version will work for Windows as well. Ignore the fact that it says linux-x64.

**Comment by mschile** [comment]
@richardszanyi-kasa, the binary I provided does include a memory fix. Would you be able to recouple one of the specs that was failing and give it a try locally with `cypress run`?

> @mschile As many of us noted previously in this thread - numTestsKeptInMemory is only a minor possible workaround..

The binary I provided includes a memory fix that impacts this as well (though I still want to make an additional change). However, depending on several factors, such as the size of your tests, the size of the app be loaded, what commands are being run, and so forth, the default value may still be too large. Thus, my recommendation to decrease the value.

**Comment by richardszanyi-kasa** [comment]
@mschile Sorry, I haven't read that, my bad. 
I'll take a look after we upgrade from 10 to 12. (It is still in progress due to a bug in @cypress/grep with typescript..)

**Comment by nickpm21** [comment]
> @richardszanyi-kasa, the binary I provided does include a memory fix. Would you be able to recouple one of the specs that was failing and give it a try locally with `cypress run`?
> 
> > @mschile As many of us noted previously in this thread - numTestsKeptInMemory is only a minor possible workaround..
> 
> The binary I provided includes a memory fix that impacts this as well (though I still want to make an additional change). However, depending on several factors, such as the size of your tests, the size of the app be loaded, what commands are being run, and so forth, the default value may still be too large. Thus, my recommendation to decrease the value.

Hi, the tests still run out of memory with your fix and reducing the number of tests kept in memory to 0.

**Comment by mschile** [comment]
@nickpm21, do you have a project I could look at?

**Comment by mschile** [comment]
@henricook, are you able to test out the binary above, https://github.com/cypress-io/cypress/issues/23391#issuecomment-1351858785?

**Comment by nickpm21** [comment]
> @nickpm21, do you have a project I could look at?

Since it's company related data I don't think I can share it, however I will mention that I rarely get out of memory errors during cypress run mode,

**Comment by mschile** [comment]
@nickpm21, with the test binary installed, could you post the result of running: `cypress info`

**Comment by nickpm21** [comment]
@mschile 
```
Displaying Cypress info...

Detected 3 browsers installed:

1. Chrome
  - Name: chrome
  - Channel: stable
  - Version: 108.0.5359.100
  - Executable: C:\Program Files\Google\Chrome\Application\chrome.exe
  - Profile: C:\Users\nickp\AppData\Roaming\Cypress\cy\production\browsers\chrome-stable

2. Chrome Beta
  - Name: chrome
  - Channel: beta
  - Version: 109.0.5414.46
  - Executable: C:\Program Files\Google\Chrome Beta\Application\chrome.exe
  - Profile: C:\Users\nickp\AppData\Roaming\Cypress\cy\production\browsers\chrome-beta

3. Edge
  - Name: edge
  - Channel: stable
  - Version: 108.0.1462.46
  - Executable: C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

Note: to run these browsers, pass <name>:<channel> to the '--browser' field

Examples:
- cypress run --browser chrome
- cypress run --browser edge

Learn More: https://on.cypress.io/launching-browsers

Proxy Settings: none detected
Environment Variables: none detected

Application Data: C:\Users\nickp\AppData\Roaming\cypress\cy\development
Browser Profiles: C:\Users\nickp\AppData\Roaming\cypress\cy\development\browsers
Binary Caches: C:\Users\nickp\AppData\Local\Cypress\Cache

Cypress Version: 12.1.0 (pre-release)
System Platform: win32 (10.0.22621)
System Memory: 34.2 GB free 23.3 GB

This is a pre-release build of Cypress.
Build info:
  Commit SHA: 5ccc7c6ae5d2a31e9625148db450b53719b1ad53
  Commit Branch: mschile/chrome_memory
  Commit Date: 2022-12-14T15:45:32.000Z
```

**Comment by nickpm21** [comment]
Since this issue has been so prevalent for so many users wouldn't it be better to run garbage collection after every test since it seems just one large file can crash the app. Is it too expensive to do this? @mschile

**Comment by mschile** [comment]
@nickpm21, there is a cost to running garbage collection. If possible, I'm planning on having a better way to determine when garbage collection needs to be run. It appears Chrome isn't doing it automatically since there isn't enough idle. You can try adding the following to your support file:

```
afterEach(() => {
  if (Cypress.isBrowser({ family: 'chromium' })) {
    Cypress.automation('remote:debugger:protocol', { command: 'HeapProfiler.collectGarbage' })
  }
})
```

**Comment by nickpm21** [comment]
Even with using collect garbage before each test it's still crashing at the same moment. It's the same thing every time. The test crashes, I reload, the test passes fine, I rerun the test and it crashes at a specific place. I'd be happy to show you personally this repeatable error if you email me. @mschile

**Comment by robrich7** [comment]
> @MusolQA, @richardszanyi-kasa, @hegeds, and any others experiencing this issue using `cypress run`, could you please try installing the below version of Cypress to see if your memory issues are fixed:
> 
> ```
> npm install https://cdn.cypress.io/beta/npm/12.1.0/linux-x64/mschile/chrome_memory-5ccc7c6ae5d2a31e9625148db450b53719b1ad53/cypress.tgz
> ```
> 
> Note: this version is based off Cypress 12.1.0

i have also tested it locally and unfortunately it still crashes on a testspec with 16 tests. i can't send you a repo either as there are company internals in there. 

`numTestsKeptInMemory = 5`

and with 

```
afterEach(() => {
  if (Cypress.isBrowser({ family: "chromium" })) {
    Cypress.automation("remote:debugger:protocol", { command: "HeapProfiler.collectGarbage" });
  }
});
```

Here is my `npx cypress info`: 

```
Detected 3 browsers installed:

1. Chrome
  - Name: chrome
  - Channel: stable
  - Version: 108.0.5359.124
  - Executable: C:\Program Files\Google\Chrome\Application\chrome.exe
  - Profile: C:\Users\RobinReicheSmartComm\AppData\Roaming\Cypress\cy\production\browsers\chrome-stable

2. Edge
  - Name: edge
  - Channel: stable
  - Version: 108.0.1462.46
  - Executable: C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

3. Firefox
  - Name: firefox
  - Channel: stable
  - Version: 108.0
  - Executable: C:\Program Files\Mozilla Firefox\firefox.exe
  - Profile: C:\Users\RobinReicheSmartComm\AppData\Roaming\Cypress\cy\production\browsers\firefox-stable

Proxy Settings: none detected
Environment Variables: none detected

Cypress Version: 12.1.0 (pre-release)  
System Platform: win32 (10.0.19045)    
System Memory: 34.2 GB free 22.1 GB    

This is a pre-release build of Cypress.
Build info:
  Commit SHA: 5ccc7c6ae5d2a31e9625148db450b53719b1ad53
  Commit Branch: mschile/chrome_memory
  Commit Date: 2022-12-14T15:45:32.000Z
```

**Comment by CoryDanielson** [comment]
> Anecdotal evidence from this morning:
> 
> **Cypress 11.2.0, Chrome 107.0.5304.110 (Official Build) (x86_64)** All of these runs were with the command log enabled.
> 
> * **crashed** `numTestsKeptInMemory: 1, no gc` OOM on the 5th test. Memory ran up to 5GB and crashed.
> * **crashed** `numTestsKeptInMemory: 1, with window.gc() calls`. Memory >2gb after 7 tests. 2.8gb after 10. OOM crash soon after
> * **suite passed** `numTestsKeptInMemory: 0, no gc` the suite of 31 tests passed. I didn't keep an eye on the memory for this run. finished at 1gb
> * **suite passed** `numTestsKeptInMemory: 0, with window.gc() calls` running great. memory gc'd down to 800mb during the 17/18th test. suite finished after 31 tests & 714mb memory.
> 
> Seems like an issue with snapshots. `numTestsKeptInMemory: 0` is a good workaround for now.

This suggests that there's a memory leak/bloat related to snapshots that is not getting cleaned up between tests, or when a tests snapshots are getting replaced/deleted.

Ironically, something like the PR that I helped author (related to css memoization) may be contributing to the memory bloat. If there is more memoization throughout the snapshot code, I think that's a good place to look and audit. Snapshots may get deleted, but memo caches may still contain references to portions of snapshots/dom nodes etc.

You may even want to disable memoization entirely to see what memory looks like. The tests will run much more slowly, but if memory is better then it'll help you isolate the problem(s).

Memory should not be running out of control with `numTestsKeptInMemory: 1` but it is. Set it to 0, and your fine. Any amount of snapshots can cause the issue on a large enough app.

--- 

The other issue that was mentioned in this thread seems like a different memory leak. Long running tests (even in headless mode) are crashing. No snapshots are made here, but it's still crashing.

Since GC isn't keeping memory under control, there must be expensive references within the program/code while it's running.

---

It's also still unknown if this was still an issue prior to Chrome 103. Nobody in the thread seems to have tried an old chrome version yet?

**Comment by CoryDanielson** [comment]
I don't see anything that manages the WeakMap used within snapshotting. My naive guess here is that this map is growing indefinitely even when snapshots are being removed between tests. I don't personally know where the code lives that deletes old snapshots between tests.

- https://github.com/cypress-io/cypress/blob/3cdbeca59d4871453d044701f2f3100cb9fbc8af/packages/driver/src/cy/snapshots.ts#L14
- https://github.com/cypress-io/cypress/blob/3cdbeca59d4871453d044701f2f3100cb9fbc8af/packages/driver/src/cy/snapshots.ts#L106
- https://github.com/cypress-io/cypress/blob/3cdbeca59d4871453d044701f2f3100cb9fbc8af/packages/driver/src/cy/snapshots.ts#L138
- https://github.com/cypress-io/cypress/blob/3cdbeca59d4871453d044701f2f3100cb9fbc8af/packages/driver/src/cy/snapshots.ts#L291
- https://github.com/cypress-io/cypress/blob/3cdbeca59d4871453d044701f2f3100cb9fbc8af/packages/driver/src/cy/snapshots.ts#L298-L304

**Comment by CoryDanielson** [comment]
It seems like within Cypress & Snapshotting, caches need to be more tightly coupled to snapshots. The CSS-related snapshot code has quite a few caches (WeakMaps, map, memoize) that don't seem to be cleaned up anywhere.

I ran a test locally and debugged it with Allocation Sampling in the Memory tab, and I was able to see large allocations for various CSS-related caches that go back to **getStyleId** or **getStyleIdsFor**.

![cy-memory](https://user-images.githubusercontent.com/1461792/208142718-d8eee064-e408-495d-a719-c102ed956120.png)
![Screen Shot 2022-12-16 at 11 32 05 AM](https://user-images.githubusercontent.com/1461792/208145995-9e00f5c8-0c02-4622-a22f-bfd8c67f300f.png)

**Comment by CoryDanielson** [comment]
Only a longer running memory allocation profile, the memory usage seems more related to the files that are apart of my application, and the files that have been loaded. The css caches are not near the top of the list.

Still curious to know if this issue is better prior to Chrome 103/104. I'll test it eventually

**Comment by JessefSpecialisterren** [comment]
@CoryDanielson wrote:
> Still curious to know if this issue is better prior to Chrome 103/104. I'll test it eventually

To comment on that: we currently run our tests using an old Chromium 99 because with that version we have no memory issues to speak of. Chrome 100 made it impossible to run a long spec in `cypress open` without getting Aw Snap, and Chrome 104 made our headless CI runs useless because it would hang every single run. Chrome 107 was usable again for CI, but tends to hang when there are failures (we have test retries enabled, so memory consumption increases when there are failures). Therefore, we switched back to Chromium 99 to ensure maximum CI stability

**Comment by hegeds** [comment]
@mschile tested it with the provided build -> there are some improvements, but I am still able to see the gradual increase during the run time.

```
09:55:57, 11315, 200M, Cypress Helper
...
10:10:08, 11315, 342M, Cypress Helper
...
10:35:32, 11315, 554M, Cypress Helper
```

What I've also found is really depends on the site under test (so when running against github the leak has a smaller slope, then when we are running it against our webapp), but in both cases the sites by themselves do not leak.

Also running it without the command log yields better results, but that is also not a full solution (tbh before commenting here I run tests, where I had gc calls in the afterEach hook)

**Comment by mschile** [comment]
@hegeds, thanks for giving the test build a try. I do expect memory to increase and then come back down. Is this what you observed? During our profiling and investigation, we haven't noticed any memory leaks of any significance but we did observe that unretained memory is not being GC'd as expected. We will be doing a much deeper dive into why GC is not happening and a more substantial profiling to ensure we are not needlessly creating objects. However, in the meantime we wanted to provide users a workable solution. The test build forces GC every 5th test which is why I would expect it to increase and then come back down.

**Comment by hegeds** [comment]
so this is plot that I got when I run the test build:
![Screenshot 2022-12-19 at 17 03 10](https://user-images.githubusercontent.com/28830478/208468384-dc436dc3-51f2-4b91-a48c-e3b5bb06c143.png)

Interestingly, when I turned off the command log I had the following result:
![Screenshot 2022-12-19 at 17 12 59](https://user-images.githubusercontent.com/28830478/208470194-d7ee01a7-f992-4d6d-8f9a-55a177038f10.png)


Could it be, that some stuff remains referenced in the command log, and that causes the memory increase? I'm just wondering, 'cause in reality there is no need for more log, than could fit in the screen at one time.

**Comment by CoryDanielson** [comment]
@hegeds Are you running `open` with snapshots enabled?

Does disabling the command log also disable snapshots, since there's no way to access snapshots without the command log?

**Comment by hegeds** [comment]
@CoryDanielson I'm running it for these results in headless mode, so `cypress run`

**Comment by mschile** [comment]
@hegeds, thanks for the great graphs! Yes, I believe the logs are still being retained during the run of the spec (though the objects in the logs are cleared out). I logged #25230 to ensure the logs get released.

**Comment by kpturner** [comment]
This is becoming a bit of a nightmare.  Up until version 12 I could at least get our tests to run through by forcing Cypress to use Chrome 99.   With version 12 not even that browser works.  I cannot get our tests to complete without OOM errors in any browser at all......which kinda rules V12 out for us for the foreseeable.

**Comment by kpturner** [comment]
@JessefSpecialisterren  wrote
> Therefore, we switched back to Chromium 99 to ensure maximum CI stability

This is becoming a bit of a nightmare. Up until version 12 I could at least get our tests to run through by forcing Cypress to use Chrome 99. With version 12 not even that browser works. I cannot get our tests to complete without OOM errors in any browser at all......which kinda rules V12 out for us for the foreseeable.

**Comment by nickpm21** [comment]
So as per previous conversations it looks like while using the GUI via "cypress open", on any browser, Chrome, Firefox, Edge, etc.. The memory used by the browser instance does not go down after finishing a test. Because this happens on multiple browsers all the same way this would seem to be pretty isolated to Cypress itself causing the issue as opposed to a Chromium version memory leak issue. Is there any way to reset memory usage for the instance while having it active? @mschile

**Comment by JessefSpecialisterren** [comment]
@kpturner wrote
> This is becoming a bit of a nightmare. Up until version 12 I could at least get our tests to run through by forcing Cypress to use Chrome 99. With version 12 not even that browser works. I cannot get our tests to complete without OOM errors in any browser at all......which kinda rules V12 out for us for the foreseeable.

Interesting! We switched to Cypress 12 last week and haven't had a single OOM problem in our CI runs with Chromium 99 so far. This issue seems to be remarkably hard to pin down 🤔

For the record: does "any browser at all" include Firefox?

**Comment by henricook** [comment]
It's only newer browsers, or at least is only particularly prominent in newer browsers. We've had to keep our CI browser tagged at around Chromium 102 to keep our tests running while the Cypress team try and figure out this problem.

**Comment by kpturner** [comment]
> @kpturner wrote
> 
> > This is becoming a bit of a nightmare. Up until version 12 I could at least get our tests to run through by forcing Cypress to use Chrome 99. With version 12 not even that browser works. I cannot get our tests to complete without OOM errors in any browser at all......which kinda rules V12 out for us for the foreseeable.
> 
> Interesting! We switched to Cypress 12 last week and haven't had a single OOM problem in our CI runs with Chromium 99 so far. This issue seems to be remarkably hard to pin down 🤔
> 
> For the record: does "any browser at all" include Firefox?

Yes it includes Firefox but I have not tried all different versions of Firefox. It’s just not practical to try every version of every browser so I can’t say all versions of FF fail in the same way.

I’m considering splitting our test suite in two as some sort of workaround - but this situation is not good.

Firefox fails in a different way: `The automation client disconnected. Cannot continue running tests.` (we are not using Docker).

**Comment by kpturner** [comment]
> It's only newer browsers, or at least is only particularly prominent in newer browsers. We've had to keep our CI browser tagged at around Chromium 102 to keep our tests running while the Cypress team try and figure out this problem.

If only that were true

**Comment by mschile** [comment]
@kpturner, could you please log a new issue for Firefox since this one pertains to Chromium browsers.

You can try forcing garbage collection in Chromium based browsers by adding the following in your support file:

```
// force garbage collection before every 5th test
let testCount = 0
beforeEach(() => {
  testCount++
  if (Cypress.isBrowser({ family: 'chromium' }) && testCount % 5 === 0) {
    Cypress.automation('remote:debugger:protocol', { command: 'HeapProfiler.collectGarbage' })
  }
})
```

**Comment by kpturner** [comment]
> @kpturner, could you please log a new issue for Firefox since this one pertains to Chromium browsers.
> 
> You can try forcing garbage collection in Chromium based browsers by adding the following in your support file:
> 
> ```
> // force garbage collection before every 5th test
> let testCount = 0
> beforeEach(() => {
>   testCount++
>   if (Cypress.isBrowser({ family: 'chromium' }) && testCount % 5 === 0) {
>     Cypress.automation('remote:debugger:protocol', { command: 'HeapProfiler.collectGarbage' })
>   }
> })
> ```

@mschile I will give that a try. However I have tests that crash chrome (using the GUI) after just two tests have run.  I am almost at the point of trying this for every test - no idea how much it will impact performance though.

My Firefox comment was just to answer the question posed to me (had I tried it).   I will log an issue later but I don't want to do so without all the baggage that goes with doing so.  At the moment I need to firefight the Chrome problem as a priority.

**Comment by mschile** [comment]
@kpturner, if you are getting OOM after a few tests in open (GUI) mode, I would also suggest setting `numTestsKeptInMemory: 0` in your Cypress config. This will prevent DOM snapshots from getting saved along with some other objects. This can be helpful for long tests or for a more memory intensive site.

**Comment by kpturner** [comment]
> @kpturner, if you are getting OOM after a few tests in open (GUI) mode, I would also suggest setting `numTestsKeptInMemory: 0` in your Cypress config. This will prevent DOM snapshots from getting saved along with some other objects. This can be helpful for long tests or for a more memory intensive site.

We have always used that setting because we have always had issues.  Unfortunately with V12 it makes no difference.  The chrome browser crashes every time we run certain tests.

**Comment by kpturner** [comment]
@mschile 
> @kpturner, could you please log a new issue for Firefox since this one pertains to Chromium browsers.
> 
> You can try forcing garbage collection in Chromium based browsers by adding the following in your support file:
> 
> ```
> // force garbage collection before every 5th test
> let testCount = 0
> beforeEach(() => {
>   testCount++
>   if (Cypress.isBrowser({ family: 'chromium' }) && testCount % 5 === 0) {
>     Cypress.automation('remote:debugger:protocol', { command: 'HeapProfiler.collectGarbage' })
>   }
> })
> ```

Tried this overnight.  Unfortunately same result

![image](https://user-images.githubusercontent.com/308646/210978631-2db17ae7-6f35-41b9-9f17-753a24078847.png)

**Comment by kpturner** [comment]
OK, doing the garbage collection **after each test** did mean I got a clean run using Chrome 99.  I also narrowed it down to a single test scenario that did a large number of `cy.screenshot` commands FWIW.    Next I will try the same run with Chrome 108.

**Comment by kpturner** [comment]
Well after much pain and trial and error it turns out that I was able to return our tests (on Chrome 99 at least) back to normal simply by setting `e2e.testIsolation` to false in the config.  Sure I had to fix a couple of breaking changes to do with aliases but no OOM errors and everything runs cleanly.  I still need to try later versions of Chrome but I suspect they will still fail.

**Comment by kpturner** [comment]
So, rather astonishingly, our OOM issues have been completely tipped on their head simply by setting `e2e.testIsolation` to `false`.  Prior to `Cypress 12` we could only get a clean run of test results using `Chrome 99`.   Now, using `Cypress 12.3.0` and setting `e2e.testIsolation` to `false`, we are able to achieve a clean run with the latest version of Chrome (108).

**Comment by ClaasWilke** [comment]
> So, rather astonishingly, our OOM issues have been completely tipped on their head simply by setting `e2e.testIsolation` to `false`. Prior to `Cypress 12` we could only get a clean run of test results using `Chrome 99`. Now, using `Cypress 12.3.0` and setting `e2e.testIsolation` to `false`, we are able to achieve a clean run with the latest version of Chrome (108).

Unfortunately, the `e2e.testIsolation` didn't help us. We are experiences OOM issues, when running our tests with both `Cypress 11.2.0` and `12.3.0`, both running with Electron (106).
Using `Cypress 10.9.0` with Electron (102) worked just fine.

Besides, this issue seems to be related to (if not a duplicate of) https://github.com/cypress-io/cypress/issues/24905

**Comment by mschile** [comment]
The code for this is done in [cypress-io/cypress#25462](https://github.com/cypress-io/cypress/pull/25462), but has yet to be released.
We'll update this issue and reference the [changelog](https://on.cypress.io/changelog) when it's released.

**Comment by kpturner** [comment]
> The code for this is done in [cypress-io/cypress#25462](https://github.com/cypress-io/cypress/pull/25462), but has yet to be released. We'll update this issue and reference the [changelog](https://on.cypress.io/changelog) when it's released.

What code are we talking about?   Has the memory leak been found and fixed or are we relying on the setting that forces garbage collection?  The latter didn't help (in our case).

**Comment by nagash77** [comment]
Hi @kpturner , during our investigation we found that Cypress was preventing chromium browsers from running their GC normally since there was no periods of inactivity.  This new approach has Cypress monitor memory usage and calling for GC when appropriate.  This does not address any potential memory leaks that may exist (we do not know of any current ones but that does not mean they can't exist).  We are planning to focus on performance and resource utilization throughout 2023, so more to come on that front in the future.  

If this new flag does not address your issue, I would ask that you please open a new ticket for your specific issue so that we can investigate that specific use case (which may have details that unearth some new findings).  Thank you.

**Comment by kpturner** [comment]
Yes we tried the experimental garbage collection step - every 5 tests and indeed on every test and Chrome still crashed.  For us, simply reverting to having `e2e.testIsolation` switched off resolved it.  

I didn't see any other feedback on here to suggest that the garbage collection option helped others.  Did I miss something? Closing the issue would suggest it is fixed for everyone else?

**Comment by hegeds** [comment]
from what I've been seeing this problem has multiple layers, one of those was gc not being able to do it's just. That is addressed in this PR.

An other thing that I found is that during long running tests that the command log is not being purged. This can be thought of a memory leak light, because it is an intended behaviour, just for a long run those entries can add up. This should be addressed in this ticket: https://github.com/cypress-io/cypress/issues/25230
